gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/store/hdfs/SimpleHDFSConfigStoreFactory.java

```diff
@@ -0,0 +1,111 @@
+   * qualified path to the dataset in question. For example,
+   * {@code simple-hdfs://[authority]:[port][path-to-config-store][path-to-dataset]}. It is important to note that the
+   * path to the config store on HDFS must also be specified. The combination
+   * {@code [path-to-config-store][path-to-dataset]} need not specify an actual {@link Path} on HDFS.
+   *
+   * @param  configKey       The URI of the config key that needs to be accessed.
+   *
+   * @return a {@link SimpleHDFSConfigStore} configured with the the given {@link URI}.
+   *
+   * @throws ConfigStoreCreationException if the {@link SimpleHDFSConfigStore} could not be created.
+   */
+  @Override
+  public SimpleHDFSConfigStore createConfigStore(URI configKey) throws ConfigStoreCreationException {
+    FileSystem fs = createFileSystem(configKey);
+    return new SimpleHDFSConfigStore(fs, getStoreRoot(fs, configKey));
+  }
+
+  /**
+   * Returns the physical scheme this {@link ConfigStoreFactory} is responsible for. To support new HDFS
+   * {@link FileSystem} implementations, subclasses should override this method.
+   */
+  protected String getPhysicalScheme() {
+    return HDFS_SCHEME_NAME;
+  }
+
+  /**
+   * Creates a {@link FileSystem} given a user specified configKey.
+   */
+  private FileSystem createFileSystem(URI configKey) throws ConfigStoreCreationException {
+    try {
+      return FileSystem.get(createHDFSURI(configKey), new Configuration());
+    } catch (IOException | URISyntaxException e) {
+      throw new ConfigStoreCreationException(configKey, e);
+    }
+  }
+
+  /**
+   * Creates a HDFS {@link URI} given a user specified configKey.
+   */
+  private URI createHDFSURI(URI configKey) throws URISyntaxException {
+    String configKeyScheme = configKey.getScheme();
+    if (!configKeyScheme.startsWith(SimpleHDFSConfigStore.SIMPLE_HDFS_SCHEME_PREFIX)) {
+      throw new IllegalArgumentException(String.format("Scheme for configKey \"%s\" must begin with \"%s\"!", configKey,
+          SimpleHDFSConfigStore.SIMPLE_HDFS_SCHEME_PREFIX));
+    }
+    return new URI(getPhysicalScheme(), configKey.getAuthority(), "" ,"", "");
+  }
+
+  /**
+   * This method determines the physical location of the {@link SimpleHDFSConfigStore} root directory on HDFS. It does
+   * this by taking the {@link URI} given by the user and back-tracing the path. It checks if each parent directory
+   * contains the folder {@link SimpleHDFSConfigStore#CONFIG_STORE_NAME}. It the assumes this {@link Path} is the root
+   * directory.
+   */
+  private URI getStoreRoot(FileSystem fs, URI configKey) throws ConfigStoreCreationException {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(configKey.getPath()),
+        String.format("Root Path must be specified for configKey: %s", configKey));
+    Path path = new Path(configKey.getPath());
+
+    while (path != null) {
+      try {
+        if (fs.exists(path)) {
+          for (FileStatus fileStatus : fs.listStatus(path)) {
```

Comment:
But that would throw a `FileNotFoundException` if `new Path(path, SimpleHDFSConfigStore.CONFIG_STORE_NAME)` does not exist, which can happen as `SimpleHDFSConfigStore.CONFIG_STORE_NAME` is only suppose to exist in one of the parent directories.